visions/accessors.md

```diff
@@ -0,0 +1,348 @@
+
+Another example:  Providing a `Span` over the contents of a String will require that we somehow handle the case where a short String is stored inline.
+
+### Access Scope
+
+The `get`/`set` accessors conceptually represent “instantaneous” access of the value.  After the `get` accessor returns, there is no relation between the returned copy of the property value and the containing value.  Each of the other accessors implies that the containing value must continue to exist for some period.  This period is generally fairly short, but optimization can extend this interval in order to simplify other operations:
+
+* `unsafe*Address` accessors return pointers into the containing value.  This is safe for the caller of these accessors because the compiler knows about this relationship and can extend the lifetime of the containing value as needed.  (These accessors are nominally “unsafe” because their implementation requires constructing an unsafe pointer and there are no checks to ensure that the pointer so constructed is in fact valid.)
+* `read`/`modify` expose a value for the lifetime of a coroutine. Coroutines enforce that the containing value remains alive for the duration of the coroutine.  Note that in current Swift, the coroutines are generally quite short-lived and the compiler copies the value into or out of the coroutine fairly aggressively.  In the future, the compiler will likely become more adept at expanding the coroutine lifetime to reduce such copying.
+* `borrow`/`mutate` return a borrow of the property value.  This borrow has an implied dependence on the containing value, and the compiler must guarantee that the containing value outlives the property access.
+
+### Ownership of the containing value
+
+All of the above accessors include a reading variant (`get`, `read`, `borrow`) and a writing variant (`set`, `modify`, `mutate`).  By default, the reading variant is not considered to be a mutation of the enclosing value.  The writing variant conversely is considered to be a write access of the enclosing value and the usual Swift exclusivity rules apply to limit the scope of such access.
+
+However, a reading accessor can be explicitly marked as `mutating`.  This relaxes checks on the accessor implementation so that it can mutate the containing value, and also causes the caller to treat this as a write access requiring exclusivity enforcement.  This is commonly used for caching or bookkeeping:
```

Comment:
Does this imply the existence of a `mutating read`? What is the difference between `mutating read` and `modify`? (I guess it gives you a read-only object while requiring the enclosing object to be mutable?)

Reply:
Exactly:  A `modify` accessor provides an API for the client to mutate the property; a `mutating read` allows the read operation to have side-effects (such as updating a cache).  Mutating reads need to be declared as such so the compiler can accurately diagnose cases where those side-effects could cause problems.